mlir/lib/Dialect/Tosa/Transforms/SinkInputOpsThroughConcat.cpp

```diff
@@ -0,0 +1,425 @@
+    }
+    auto opName = sample->getName().getStringRef();
+    auto amount = ++stats[opName];
+    if (amount == 1 || amount % 50 == 0)
+      os << "SinkInputOpsThroughConcat: Operation " << opName
+         << " -> Matched amount: " << amount << "\n";
+    return rewriter.notifyMatchFailure(concatOp,
+                                       "Only for statistics printing");
+  }
+
+private:
+  llvm::StringMap<unsigned> &stats;
+  llvm::raw_ostream &os;
+};
+
+struct CheckResult {
+  bool rewriteIsPossible;
+  const char *msg;
+};
+
+template <class OpT>
+struct GenericCheckResult {
+  OpT operand;
+  const char *msg;
+};
+
+template <class OpT>
+bool checkForEquality(OpT sample, OpT op) {
+  for (auto attrName : OpT::getAttributeNames()) {
+    if (sample->getAttr(attrName) != op->getAttr(attrName)) {
+      return false;
+    }
+  }
+  return llvm::equal(op->getOperandTypes(), sample->getOperandTypes());
+}
+
+template <class OpT>
+struct SinkSpecificOp : public OpRewritePattern<tosa::ConcatOp> {
+  using OpRewritePattern<tosa::ConcatOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(tosa::ConcatOp concatOp,
+                                PatternRewriter &rewriter) const override {
+    auto genChecks = doGenericChecks(concatOp);
+    auto sample = genChecks.operand;
+    if (!sample) {
+      return rewriter.notifyMatchFailure(concatOp, genChecks.msg);
+    }
+
+    const auto extraChecks = this->doExtraChecks(sample, concatOp);
+    if (!extraChecks.rewriteIsPossible) {
+      return rewriter.notifyMatchFailure(concatOp, extraChecks.msg);
+    }
+
+    // rewriting, first collect all inputs and rearrange them for the concats
+    SmallVector<SmallVector<Value>> concatOperands(sample->getNumOperands());
+    for (auto val : concatOp.getOperands()) {
+      auto *rewriteOp = val.getDefiningOp();
+      assert(rewriteOp != nullptr &&
+             "Previous check about null already happened.");
+      for (unsigned j = 0; j < rewriteOp->getNumOperands(); ++j) {
+        if (concatOperands.size() <= j) {
+          concatOperands.emplace_back(
+              SmallVector<Value>(concatOp->getNumOperands()));
+        }
+        concatOperands[j].emplace_back(rewriteOp->getOperand(j));
+      }
+    }
+
+    // then create the concats and replacement op
+    SmallVector<Value> concatOps;
+    for (auto ops : concatOperands) {
+      auto concatOpReplacement = rewriter.create<tosa::ConcatOp>(
+          concatOp.getLoc(), ops, concatOp.getAxis());
+      concatOps.emplace_back(concatOpReplacement);
+    }
+    // leave it generic here to avoid a bunch of different constructors
+    auto *switchedOp =
+        rewriter.create(concatOp->getLoc(), sample->getName().getIdentifier(),
+                        concatOps, concatOp.getType());
+    for (auto attr : sample->getAttrs()) {
+      switchedOp->setAttr(attr.getName(), attr.getValue());
+    }
+    rewriter.replaceOp(concatOp, switchedOp);
```

Comment:
I think you can simplify this:
```suggestion
    rewriter.replaceOpWithNewOp<OpT>(concatOp, concatOp.getResultTypes(), concatOps, sample->getAttrs());
```

Reply:
That is much better! Thanks.